airbyte-integrations/connectors/source-orb/source_orb/source.py

```diff
@@ -177,6 +180,261 @@ def transform_record(self, subscription_record):
+
+        # self.logger.warning("new request_params are: %s", params)
+        return params
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        # self.logger.warning("calling get_updated_state with current_stream_state %s and latest_record %s", current_stream_state, latest_record)
+        current_stream_state = current_stream_state or {}
+        current_subscription_id = latest_record["subscription_id"]
+        current_subscription_state = current_stream_state.get(current_subscription_id, {})
+
+        record_cursor_value = to_datetime(latest_record[self.cursor_field])
+        state_cursor_value = to_datetime(current_subscription_state.get(self.cursor_field, self.start_date))
+        max_cursor_value = max(record_cursor_value, state_cursor_value)
+
+        current_subscription_state[self.cursor_field] = to_utc_isoformat(max_cursor_value)
+
+        return {
+            **current_stream_state,
+            current_subscription_id: current_subscription_state
+        }
+
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs):
+        """
+        Orb does not support querying for all subscription usage in an unscoped
+        way, so the path here is dependent on the stream_slice, which determines
+        the `subscription_id`.
+        """
+        subscription_id = stream_slice["subscription_id"]
+        return f"subscriptions/{subscription_id}/usage"
+
+    def get_billable_metric_ids_by_plan_id(self) -> Mapping[str, Any]:
+        metric_ids_by_plan_id = {}
+
+        for plan in Plans(authenticator=self._session.auth).read_records(sync_mode=SyncMode.full_refresh):
+            # if a plan_id filter is specified, skip any plan that doesn't match
+            if self.plan_id and plan["id"] != self.plan_id:
+                continue
+
+            prices = plan.get("prices", [])
+            metric_ids_by_plan_id[plan["id"]] = [(price.get("billable_metric") or {}).get("id") for price in prices]
+
+        # self.logger.warning("returning %s from get_billable_metric_ids", metric_ids_by_plan_id)
+        return metric_ids_by_plan_id
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        This schema differs from `subscription_usage.json` based on the configuration
+        of the Stream. If a group_by key is specified, the stream will output
+        records that contain the group_key name and value.
+        """
+        schema = super().get_json_schema()
+        if self.subscription_usage_grouping_key:
+            schema[self.subscription_usage_grouping_key] = {"type": "string"}
+
+        return schema
+
+    def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, Any]]]:
+        """
+        This stream is sliced per `subscription_id` and day, as well as `billable_metric_id`
```

Comment:
the logic feels complex because the stream is sliced on 3 dimensions. That's a fun!